### PR TITLE
Optimised db read and write 

### LINKED ISSUE #459: DB access optimizations

### PR DESCRIPTION
**Pull request by ashi20192020ayush**
resolves #459

### CODE REVIEW

File: pallets/amm/src/lib.rs [hunk 1]
```diff
@@ -155,107 +155,123 @@ pub mod pallet {
...
             };
 
-            if Pools::<T, I>::contains_key(base_asset, quote_asset) {
-                let pool_liquidity_amount: PoolLiquidityAmount =
-                    Self::pools(base_asset, quote_asset);
+            Pools::<T, I>::try_mutate(
+                &base_asset,
+                &quote_asset,
+                |pool_liquidity_amount| -> DispatchResultWithPostInfo {
+                    if let Some(liquidity_amount) = pool_liquidity_amount {
```
Comment (GopherJ) [inline pallets/amm/src/lib.rs:162]: let's avoid nested if scope here, maybe you can consider handling `None` case first and then quick return

File: pallets/amm/src/lib.rs [hunk 2]
```diff
@@ -155,107 +155,123 @@ pub mod pallet {
...
-                    } else {
-                        let optimal_base_amount = Self::quote(
-                            quote_amount,
-                            pool_liquidity_amount.quote_amount,
-                            pool_liquidity_amount.base_amount,
+                        let (base_amount, quote_amount) = (ideal_base_amount, ideal_quote_amount);
+                        let ownership = sp_std::cmp::min(
+                            (base_amount.saturating_mul(liquidity_amount.ownership))
+                                .checked_div(liquidity_amount.base_amount)
+                                .expect("cannot overflow with positive divisor; qed"),
```
Comment (GopherJ) [inline pallets/amm/src/lib.rs:199]: maybe it should fail and just revert back storage instead?

File: pallets/amm/src/lib.rs [hunk 3]
```diff
@@ -155,107 +155,123 @@ pub mod pallet {
...
-
-                Self::deposit_event(Event::<T, I>::LiquidityAdded(who, base_asset, quote_asset));
-                Ok(Some(T::WeightInfo::add_liquidity_existing_pool()).into())
-            }
+                        liquidity_amount.ownership = ownership;
+
+                        *pool_liquidity_amount = Some(liquidity_amount.clone());
+
+                        LiquidityProviders::<T, I>::try_mutate(
+                            (who.clone(), base_asset, quote_asset),
```
Comment (GopherJ) [inline pallets/amm/src/lib.rs:218]: can this `clone` be removed?

File: pallets/amm/src/lib.rs [hunk 4]
```diff
@@ -273,77 +289,73 @@ pub mod pallet {
...
+                        .ok_or(Error::<T, I>::PoolDoesNotExist)?;
+                    ensure!(
+                        liquidity_amount.ownership >= ownership_to_remove,
+                        Error::<T, I>::MoreLiquidity
+                    );
+
+                    let base_amount = (ownership_to_remove
+                        .saturating_mul(liquidity_amount.base_amount))
+                    .checked_div(liquidity_amount.ownership)
+                    .expect("cannot overflow with positive divisor; qed");
```
Comment (GopherJ) [inline pallets/amm/src/lib.rs:307]: shall this expect be replaced and just revert storage changes?

File: pallets/amm/src/lib.rs [hunk 5]
```diff
@@ -273,77 +289,73 @@ pub mod pallet {
...
                         .ownership
                         .checked_sub(ownership_to_remove)
                         .ok_or(ArithmeticError::Underflow)?;
-                    Ok(())
-                },
-            )?;
+                    *pool_liquidity_amount = Some(liquidity_amount.clone());
+
+                    LiquidityProviders::<T, I>::try_mutate(
+                        (who.clone(), base_asset, quote_asset),
```
Comment (GopherJ) [inline pallets/amm/src/lib.rs:329]: can this clone be removed?

File: pallets/amm/src/weights.rs [hunk 6]
```diff
@@ -35,58 +35,56 @@
...
-    }
-    fn add_liquidity_existing_pool() -> Weight {
-        (84_000_000 as Weight)
-            .saturating_add(T::DbWeight::get().reads(6 as Weight))
-            .saturating_add(T::DbWeight::get().writes(6 as Weight))
-    }
-    fn remove_liquidity() -> Weight {
-        (83_000_000 as Weight)
-            .saturating_add(T::DbWeight::get().reads(7 as Weight))
-            .saturating_add(T::DbWeight::get().writes(6 as Weight))
```
Comment (GopherJ) [inline pallets/amm/src/weights.rs:71]: maybe we should define `rustfmt.toml`

File: pallets/amm/src/lib.rs (hunk 1 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/lib.rs:162]: @GopherJ I have to handle both cases here a) When there is already a pool with same currency set b) When there is no pool

cc @ETeissonniere

File: pallets/amm/src/lib.rs (hunk 2 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/lib.rs:199]: @GopherJ Could you please refer any example or hint?

File: pallets/amm/src/lib.rs (hunk 3 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/lib.rs:218]: Its needed here. It can't be removed.

File: pallets/amm/src/lib.rs (hunk 5 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/lib.rs:329]: Its needed here. It can't be removed.

File: pallets/amm/src/lib.rs (hunk 2 above)
Comment (GopherJ) [reply pallets/amm/src/lib.rs:199]: I mean you are using `try_mutate` and it accepts `DispatchResult..` maybe `.ok_or(Ari...::Underflow)?` like in loans pallet

Comment (GopherJ) [reply pallets/amm/src/lib.rs:199]: we don't want `expect` and `unwrap`

File: pallets/amm/src/weights.rs (hunk 6 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/weights.rs:71]: I am running `cargo fmt` before pushing the code.

File: pallets/amm/src/lib.rs (hunk 4 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/lib.rs:307]: Could you please share any example?

File: pallets/amm/src/lib.rs (hunk 5 above)
Comment (GopherJ) [reply pallets/amm/src/lib.rs:329]: wont `&who` be enough?

**Review by ETeissonniere — APPROVED** [review]

File: pallets/amm/src/lib.rs (hunk 5 above)
Comment (ashi20192020ayush) [reply pallets/amm/src/lib.rs]: Yes @GopherJ this would work. I have modified. Please review

File: pallets/amm/src/lib.rs (hunk 2 above)
Comment (ayushmishra2005) [reply pallets/amm/src/lib.rs]: Done

File: pallets/amm/src/lib.rs (hunk 3 above)
Comment (ayushmishra2005) [reply pallets/amm/src/lib.rs]: done

File: pallets/amm/src/lib.rs (hunk 4 above)
Comment (ayushmishra2005) [reply pallets/amm/src/lib.rs]: done

**Review by GopherJ — APPROVED** [review]

**Pull request merged**